File: EAND Python/Maximization problem/Non-integer variables/eand_limit.py
''' 
Implementation of Evolutionary Algorithm with Numerical differentiation (EAND) in Python
 
Article: Soft Sensors Design in a Petrochemical Process using an Evolutionary Algorithm 
Journal: Measurement
Year: 2019

Developer: Gustavo A. P. de Morais
Contact Info: gustavo_gapm@usp.br, gustavo_gapm@hotmail.com

Constraint handling procedure applied after the mutation procedure
'''

import numpy as np

def eand_limit(M,LB,UB) :    
    
    for i in range(np.size(M)) :
        if M[i] > UB :
            M[i] = np.random.uniform(.75*UB, UB, 1)
        elif M[i] < LB :
            M[i] = np.random.uniform(0, .25*UB, 1)  
       
    return M

File: EAND MATLAB/Maximization problem/Integer variables/eand_limit.m
% 
% Implementation of Evolutionary Algorithm with Numerical differentiation (EAND) in MATLAB
% 
% Article: Soft Sensors Design in a Petrochemical Process using an Evolutionary Algorithm 
% Journal: Measurement
% Year: 2019
%
% Developer: Gustavo A. P. de Morais
% Contact Info: gustavo_gapm@usp.br, gustavo_gapm@hotmail.com
%
% Constraint handling procedure applied after the mutation procedure

function [L] = eand_limit(L,VarMin,VarMax) 

    Aux = size(L);  
    for j = 1:Aux(2) 
        if L(1,j) > VarMax 
            aux = round(unifrnd(0,0.25*(abs(VarMax)),1));
            L(1,j) = VarMax - aux;
        elseif L(1,j) < VarMin
            aux = round(unifrnd(0,0.25*(abs(VarMax)),1));
            L(1,j) = VarMin + aux;
        end
    end
           
end